### PR TITLE
feat: added cors middleware

### PR DESCRIPTION
**Pull request by prisis**
## Summary

A CORS middleware.

## TODO

- [x] add tests
- [x] update readme

### CODE REVIEW

**Comment by ebebbington** [comment]
Hi @prisis, i'm still looking through it but looking forward to it! 😄 , but I think the approach we will take to middleware is to export the middleware from it's own `mod.ts`, for example:
```
deno-drash-middleware/
    src/
        cors/
            your_cors_file.ts
            mod.ts
        another_middleware/
            another_middleware_file.ts
            mod.ts
```
And it would be imported like: `import { Cors } from "https://deno.land/deno-drash-middleware@v1.0.0/cors/mod.ts`

I don't know what the approach around importing 3rd party modules is yet, because it can get really messy, i believe it's best (for the modules you need) to have your own `deps.ts` in your `cors` directory - and that will be all the 3rd party dependencies you need, and the std deps will still be in the root dependency file. I say this because it means if one dependency breaks with Deno, every middleware will break (as its all imported/exported from the same directory). Though i'm not entirely sure on this. Would need clarification by @crookse for this point

**Comment by ebebbington** [comment]
> Do you have a structure in mind for the tests, or for the hole repository?

Feel free to create your tests in `deno-drash-middleware/tests/cors/` :) Sorry about not having all the info right away, this repo is very new (as you know) and we're a bit stuck for time at the moment

File: src/cors/cors_middleware.ts
Comment (ebebbington) [inline src/cors/cors_middleware.ts:6]: Not a problem here, but we plan on creating an interface for Drash's `request`, so if this PR gets merged ill make an issue for someone (you ideally :D) to update that type when we create Drash's request interface

**Comment by ebebbington** [comment]
LGTM at the moment, that's some really clean code!

**Comment by crookse** [comment]
So the filesystem structure will look like this:

```
deno-drash-middleware/
    cors/
        /tests
        deps.ts
        some_related_file.ts
        another_related_file.ts
        mod.ts
    another_middleware/
        /tests
        deps.ts
        some_related_file.ts
        another_related_file.ts
        mod.ts
    deps.ts
```

There's no need for a `src` directory in the root of the project. The project will be a collection of directories--each being a separate middleware module.

Tests should be written in the the directory relative to them. So you'll have a tests directory in your `cors` directory.

Third-party modules are OK, but please place third-party modules in your middleware directory's `deps.ts` file.

Any Deno std module that's required by your package can be exported from the repo's root `deps.ts` file so that all middleware can import std modules from that single file.

File: deps.ts
Comment (crookse) [inline deps.ts:1]: Drash and ServerRequest can be exported from this repo's root deps.ts file. I know there isn't one yet, so you'll be creator of it :). We don't want the potential to have middleware that uses different Drash versions, so that's why it should go in the root deps.ts file. When we bump Drash's version, we'll make sure all middleware works for the same Drash version. If a middleware doesn't work, then we'll have to debug before we make a new release.

Also, this could change in the future. This is just what we're going with right now and we'll work out any issues with this workflow as time goes on.

File: deps.ts
Comment (crookse) [inline deps.ts:5]: Place vary in your middleware directory's deps.ts file. We're going to keep third-party modules separated from the root deps.ts file.

File: mod.ts
Comment (crookse) [inline mod.ts:1]: this file isn't needed. users will have to `import { CorsMiddleware } from "https://deno.land/x/drash-middleware@v{version}/cors/mod.ts"`;

File: src/cors/cors_middleware.ts
Comment (prisis) [reply src/cors/cors_middleware.ts:6]: I can use ServerRequest from Deno for now, if its okay

Comment (crookse) [reply src/cors/cors_middleware.ts:6]: Does ServerRequest work when you use this middleware with Drash? I’m afraid there might be some weird side effects if it’s typed as ServerRequest because we use any. Would love to see tests exercise the implementation of this.

Comment (prisis) [reply src/cors/cors_middleware.ts:6]: im using the middleware in a deno app, without problems and using ServerRequest as type

**Comment by prisis** [comment]
if #3 is merged i will continue the work on this pr

**Comment by Guergeiro** [comment]
Hello @prisis. We are working on Drash Request (see it [here](https://github.com/drashland/deno-drash/pull/274)). After that is merged, I feel like your cors feature is ready to be shipped (you can change the ServerRequest to Drash Request). Just need to change the project structure. Are you okay with that?

**Comment by prisis** [comment]
Do you know why the test is not working?

**Comment by crookse** [comment]
> Do you know why the test is not working?

You need to add Rhum.run() at the end

**Comment by alex-arriaga** [comment]
Hi, first of all, thank you for your great work! I wonder if there is any news about making this CORS middleware available for usage in Drash projects soon?

**Comment by ebebbington** [comment]
Thanks @alex-arriaga, and i'm. looking forward to this middleware as well - all thanks to @prisis 

I think @prisis is pretty busy, but i'll let him speak for himself - @prisis it seems like the main logic is done with, just a few bits and bobs left for the tests, did you want me to finish it off? (eg the boring stuff :P)

**Comment by prisis** [comment]
@ebebbington, yeah its bit busy on my side, i start to write tests, but i had/have some issue with the content-length  
  
But its are the tests that i wanted to port https://github.com/fastify/fastify-cors/blob/master/test/index.test.js  
  
@ebebbington if you want to write the test, go on \<3, you have a better knowledge about drash

File: cors/deps.ts
Comment (prisis) [inline cors/deps.ts:1]: I need to release this package, reminder for me

**Comment by ebebbington** [comment]
@prisis I don't mind, it can be pretty boring not very interesting to do so I don't mind helping out - if/when i have the time myself i'll try look into it if you haven't already tackled it - of course it'd still be 100% your work of course ;)

**Comment by stephaned-ev** [comment]
Hello guys, do you have any idea when this feature will be released?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:168]: @prisis What does the `OPTIONS` do here? Would users send an `OPTIONS` request?

**Comment by ebebbington** [comment]
@stephaned-ev hey, it's hard to tell at the moment, i'm currently trying to help @prisis out by finishing off this PR

**Comment by ebebbington** [comment]
Still the issue of the tests. failing which i cannot understand why yet

**Comment by ebebbington** [comment]
Main changes from the commit above were to just improve the testing file(s), because the test file structure has improved a lot since this PR was opened

**Comment by ebebbington** [comment]
Found the issue, tests fail but will pass once https://github.com/drashland/deno-drash/pull/401 is merged. So on hold until then

**Comment by ebebbington** [comment]
Gonna tackle this again now drash's new release holds the fix

**Comment by ebebbington** [comment]
Nevermind, the release with the fix isn't released yet.

This PR can be tackled again once drash v1.3.0 is released, and when it is, the `deps.ts` in this branch should be updated, and then the tests should pass

File: cors/README.md
Comment (ebebbington) [inline cors/README.md:12]: Note to self: i need to change this then add this to the bumper process

File: cors/README.md
Comment (ebebbington) [inline cors/README.md:20]: Same as above

**Review by ebebbington — CHANGES_REQUESTED** [review]

**Review by ebebbington — COMMENTED** [review]
Apart from my cleanup commits, LGTM, thanks and nice job @prisis!

Can someone from the drashland team also review this too? @crookse i believe you're reviewing this PR too?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:61]: Can this be typed?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:11]: Should the request and response be a part of this type?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:140]: Can this variable be named so it's descriptive of it's purpose?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:111]: Can this variable be named so it's descriptive of it's purpose?

File: cors/mod.ts
Comment (ebebbington) [inline cors/mod.ts:194]: Could this be typed to be `allowHeaders`?

**Review by ebebbington — CHANGES_REQUESTED** [review]

File: cors/mod.ts
Comment (ebebbington) [reply cors/mod.ts]: https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b

File: cors/mod.ts
Comment (ebebbington) [reply cors/mod.ts]: https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b

File: cors/mod.ts
Comment (ebebbington) [reply cors/mod.ts]: https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b

File: cors/mod.ts
Comment (ebebbington) [reply cors/mod.ts]: https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b

File: cors/mod.ts
Comment (ebebbington) [reply cors/mod.ts]: https://github.com/drashland/deno-drash-middleware/pull/1/commits/1627c064bf674eeeccbe726cd79bd685da78dd8b

**Comment by ebebbington** [comment]
I think the final thing to do before we can merge this, is to have a concrete example/test of CORS allowing a request, where without the middleware it didn't allow the request

Does anyone know enough about cors to turn this into a test? Or if knowledge of Rhum/deno isn't well, you can provide the steps and i can turn that into code for a test

**Comment by crookse** [comment]
i'm going to rewrite the README so it's easily digestible and i'm going to put it in tutorial format. just mentioning it here so @prisis isn't surprised by the change after this merges. i'm very picky about our tutorials :) btw you have provided a great head start on the documentation prisis. thank you for that!

**Comment by ebebbington** [comment]
also, judging from express' cors middleware tests, their tests to show it works is what we have, i just feel there needs to be a concrete test between two servers maybe?

**Comment by crookse** [comment]
@ebebbington, i agree. if we can get a test to show that two servers can/can't communicate then let's do it. otherwise, let's just mock requests to a server with and without cors enabled

**Comment by ebebbington** [comment]
I know how to do it, gonna work on it now so we can finally pump out the cors middleware

**Comment by ebebbington** [comment]
Can't figure out how to replicate a cors issue (and using cors to fix it) using two servers, but I have tested the middleware myself and using Cors DOES work as expected

Could someone else review this as well please?

**Comment by ebebbington** [comment]
Going to try get a test implemented before merging, to make sure the cors middleware will allow a request.

My current thinking is:
```
1. create a server without cors (using 127.0.0.1 as the hostname)
2. send a fetch request from origin 127.0.0.1, to localhost
3. cors error
4. create a server with cors
5. same step as 2
6. no cors error
```

I have a manual test working, which is:
```
a script that does:

1. create server 1 using hostname 127.0.0.1 and port 1667 with a home resource
2. create server 2 using same hostname and port 1668 with a home resource
3. run servers
4. go to 127.0.0.1:1667
5. send a fetch request to http://localhost:1668
6. see cors error
7. add cors middleware to server 2
8. repeat steps 4 and 5
9. see no cors error (and the fetch res has `cors` as the type
```

My test file:
```typescript
import { Drash } from "https://deno.land/x/drash/mod.ts";
import { Cors } from "./cors/mod.ts";

class Res extends Drash.Http.Resource {
  static paths = ["/"];
  public GET() {
    this.response.body = "hello";
    return this.response;
  }
}

const server1 = new Drash.Http.Server({
  middleware: {
    after_request: [Cors()],
  },
  resources: [Res],
});

const server2 = new Drash.Http.Server({
  // middleware: {
  //   after_request: [Cors()],
  // },
  resources: [Res],
});

await server1.run({
  hostname: "127.0.0.1",
  port: 1667,
});

await server2.run({
  hostname: "127.0.0.1",
  port: 1668,
});

console.log("done");
```

**Comment by ebebbington** [comment]
Tried, and failed... think cors is implemented at the browser level.

Although not a proper example, testing if the `access-control-allow-origin:` header is present us the best bet

**Comment by ebebbington** [comment]
Added another test, to try show even more that the cors middleware works

Ready for review 👍

**Comment by ebebbington** [comment]
Updated readme, DEFINITELY ready for review now

**Pull request merged**
